docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,232 @@
+
+`Program.cs` (Pre-generated code)
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
+/* D */ [GeneratedDllImportAttribute]
+/* B */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C*/ QueryPerformanceCounter(out count);
+```
+
+Observe point (D), the new attribute. This attribute provides an indication to a Source Generator that the following declaration should be updated.
+
+`Program.cs` (In-memory updated code)
+
+``` CSharp
+/* D */ [GeneratedDllImportAttribute]
+/* B */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C */ QueryPerformanceCounter(out count);
+```
+
+During the source generation process the `DllImportAttribute` (A) would be removed. The `GeneratedDllImportAttribute` (D) would remain to provide a metadata indication that the function was auto-generated. Also note that the method declaration has been updated and marked `partial`. The Source Generator would then generate the source for this partial method. The invocation (C) remains unchanged.
+
+`Stubs.g.cs`:
+
+``` CSharp
+/* E */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount)
+{
+    unsafe
+    {
+        long result = 0;
+        bool success = QueryPerformanceCounter(&result) != 0;
+        lpPerformanceCount = result;
+        return success;
+    }
+}
+
+[DllImportAttribute("Kernel32.dll")]
+/* F */ private static extern int QueryPerformanceCounter(long* lpPerformanceCount);
+```
+
+The Source Generator would generate the implementation of the partial method (E) in a separate translation unit (`Stubs.g.cs`). At point (F) the `DllImportAttribute` declaration from the user's original declaration is copied into a private P/Invoke specifically for the generated code. The P/Invoke signature from the original declaration would be modified to contain only [blittable types][blittable_link] to ensure the JIT could inline the invocation. Finally note that the user's original function signature would remain in to avoid impacting existing callsites.
+
+In this system it is not defined how marshaling of specific types would be performed. The built-in runtime has complex rules for some types, and it is these rules that once shipped become the de facto standard - often times regardless if the behavior is a bug or not. The design here is not concerned with how the arguments go from a managed to unmanaged environment. With the IL Stub generation extracted from the runtime new type marshaling (e.g. `Span<T>`) could be introduced without requiring an corresponding update to the runtime itself. The `Span<T>` type is good example of a type that at present has no support for marshaling, but with Source Generators, users could update to the latest generator and have support without changing the runtime.
```

Comment:
A few questions related to marshallers:
1. In this example, isn't an ILstub still being generated?  Is there a way to avoid that stub, or have it pre-generated?
2. Will stubs.g.cs be human writable?  If so, how?  
3. How would a new marshaller, like Span<T>, be introduced?  How do we envision adding them to our project?  (Nuget?)
4. What is the diagnostics experience if I use a type that is not supported by the source generator?  A build error?  An exception emited for runtime?  (eg. Trying Span<T> before such a marshaller is introduced)

Reply:
@jeffschwMSFT Here are my thoughts:

> 2. Will stubs.g.cs be human writable?  If so, how?

Sure. That is already possible since users can convert their non-blittable P/Invoke signature into a blittable one and instead implement a wrapper that does all the marshaling manually. There would be little need to do any generation if the signature only contain blittable types - at least that seems to be the tenor of the current thought.

> 3. How would a new marshaller, like Span, be introduced?  How do we envision adding them to our project?  (Nuget?)

That would be a design consideration for the default Source Generator supplied by the .NET team. If that source generator is extensible or not is entirely up to us. Third parties that want their own P/Invoke source generator would need to define their own mechanism to extensible marshaling. For the default, we would add support as types become interesting. `Span<T>` would of course be in the default as well as `Nullable<T>` and various others as the language continues to evolve, but we could add support and users could opt in without requiring a runtime update.

> 4. What is the diagnostics experience if I use a type that is not supported by the source generator?  A build error?  An exception emited for runtime?  (eg. Trying Span before such a marshaller is introduced)

That is an interesting question for the Source Generator team. @chsienki How is the Source Generator team thinking about propagation of source generator states? How does a Source Generator convey failures to the users?

Reply:
Generators can produce diagnostics (much like an analyzer does today), which are presented to the user via the command line / IDE etc.

Reply:
Thanks @chsienki. @jeffschwMSFT Seems like any Source Generator for P/Invokes would be able to alert a user when a type is hit that can't be marshaled in a nice message. This could point to the classic P/Invoke mechanism when the Source Generator can't support the scenario and/or we point users to a repo to file an issue or guidance for how to manually work around support.